gcloud/rest/auth/token.py

```diff
@@ -1,95 +1,199 @@
+"""Google Cloud auth via service account file"""
 import datetime
+import enum
 import json
-import logging
 import os
 import threading
 import time
+from typing import Any
+from typing import Dict
+from typing import List  # pylint: disable=unused-import
+from typing import Optional
+
 try:
     from urllib.parse import urlencode
+    from urllib.parse import quote_plus
 except ImportError:
     from urllib import urlencode
+    from urllib import quote_plus
 
-# N.B. the cryptography library is required when calling jwt.encrypt() with
-# algorithm='RS256'. It does not need to be imported here, but this allows us
-# to throw this error at load time rather than lazily during normal operations,
-# where plumbing this error through will require several changes to otherwise-
-# good error handling.
-import cryptography  # pylint: disable=unused-import
+import backoff
 import jwt
 import requests
 
 
-TOKEN_URI = 'https://accounts.google.com/o/oauth2/token'
-TIMEOUT = 60
-
-log = logging.getLogger(__name__)
+GCE_METADATA_BASE = 'http://metadata.google.internal/computeMetadata/v1'
+GCE_METADATA_HEADERS = {'metadata-flavor': 'Google'}
+GCE_ENDPOINT_PROJECT = '{}/project/project-id'.format(GCE_METADATA_BASE)
+GCE_ENDPOINT_TOKEN = \
+    '{}/instance/service-accounts/default/token?recursive=true'\
+    .format(GCE_METADATA_BASE)
+GCLOUD_TOKEN_DURATION = 3600
+REFRESH_HEADERS = {'Content-Type': 'application/x-www-form-urlencoded'}
+
+
+class Type(enum.Enum):
+    AUTHORIZED_USER = 'authorized_user'
+    GCE_METADATA = 'gce_metadata'
+    SERVICE_ACCOUNT = 'service_account'
+
+
+def get_service_data(service):
+    # type: (Optional[str]) -> Dict[str, Any]
+    service = service or os.environ.get('GOOGLE_APPLICATION_CREDENTIALS')
+    if not service:
+        cloudsdk_config = os.environ.get('CLOUDSDK_CONFIG')
+        sdkpath = (cloudsdk_config
+                   or os.path.join(os.path.expanduser('~'), '.config',
+                                   'gcloud'))
+        service = os.path.join(sdkpath, 'application_default_credentials.json')
+        set_explicitly = bool(cloudsdk_config)
+    else:
+        set_explicitly = True
+
+    try:
+        with open(service, 'r') as f:
+            data = json.loads(f.read())
+            return data
+    except FileNotFoundError:
```

Comment:
`FileNotFoundError` isn't defined in Py2 which raises `IOError` on a file that wasn't found:
```python
>>> open('foobar', 'r')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
IOError: [Errno 2] No such file or directory: 'foobar'
```

`FileNotFoundError` inherits from `OSError`, so I think you'll have to check for both `IOError` and `OSError` to make this portable.

Reply:
yes, I realized that too after the lint/hook failures. Have a change in place to fix this in the next update to this PR.